lib/dry/component/loader.rb

```diff
@@ -3,33 +3,56 @@
 module Dry
   module Component
     class Loader
-      IDENTIFIER_SEPARATOR = '.'.freeze
-      PATH_SEPARATOR = '/'.freeze
+      class Component
+        attr_reader :loader
+        attr_reader :identifier, :path, :file
 
-      attr_reader :identifier, :path, :file
+        def initialize(loader, input)
+          @loader = loader
 
-      def initialize(input)
-        @identifier = input.to_s.gsub(PATH_SEPARATOR, IDENTIFIER_SEPARATOR)
-        @path = input.to_s.gsub(IDENTIFIER_SEPARATOR, PATH_SEPARATOR)
-        @file = "#{path}.rb"
-      end
+          @identifier = input.to_s.gsub(loader.path_separator, loader.namespace_separator)
+          if loader.namespace
+            re = /^#{Regexp.escape(loader.namespace)}#{Regexp.escape(loader.namespace_separator)}/
+            @identifier = @identifier.gsub(re, '')
+          end
 
-      def namespaces
-        identifier.split(IDENTIFIER_SEPARATOR).map(&:to_sym)
-      end
+          @path = input.to_s.gsub(loader.namespace_separator, loader.path_separator)
+          @file = "#{path}.rb"
+        end
 
-      def constant
-        Inflecto.constantize(name)
-      end
+        def namespaces
+          identifier.split(loader.namespace_separator).map(&:to_sym)
+        end
+
+        def constant
+          Inflecto.constantize(constant_name)
+        end
+
+        def instance(*args)
+          constant.new(*args)
+        end
 
-      def instance(*args)
-        constant.new(*args)
+        private
+
+        def constant_name
+          Inflecto.camelize(path)
+        end
       end
 
-      private
+      PATH_SEPARATOR = '/'.freeze
+
+      attr_reader :namespace
+      attr_reader :namespace_separator
+      attr_reader :path_separator
+
+      def initialize(container)
+        @namespace = container.config.namespace
+        @namespace_separator = container.config.namespace_separator
+        @path_separator = PATH_SEPARATOR
+      end
```

Comment:
That'd work too. I suppose we could actually have a loader-specific config namespace that we pass so only appropriate config gets sent along. Seem better?

> On 12 Jul 2016, at 6:00 PM, Piotr Solnica <notifications@github.com> wrote:
> 
> In lib/dry/component/loader.rb:
> 
> >        end
> >  
> > -      private
> > +      PATH_SEPARATOR = '/'.freeze
> > +
> > +      attr_reader :namespace
> > +      attr_reader :namespace_separator
> > +      attr_reader :path_separator
> > +
> > +      def initialize(container)
> > +        @namespace = container.config.namespace
> > +        @namespace_separator = container.config.namespace_separator
> > +        @path_separator = PATH_SEPARATOR
> > +      end
> It doesn't seem to be a good idea to couple Loader with an entire container like that. Why not just pass in the config?
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>